Simplify amortized cleanup algorighm for weak hash structures

achristensen07 opened the pull request:
#### 4d7698241f39bd3d19ba4cddad79118149a502ee
<pre>
Simplify amortized cleanup algorighm for weak hash structures
<a href="https://bugs.webkit.org/show_bug.cgi?id=259847">https://bugs.webkit.org/show_bug.cgi?id=259847</a>
rdar://113419306

Reviewed by Chris Dumez.

In 266594@main Chris fixed an important bug where if you only add to a
ThreadSafeHashSet, you never hit the cleanup condition of the container.
Other containers had this same issue.  This simplifies the algorithm used
to determine whether to clean up the container to just this:

1. Increase the operation count when an add/remove operation happens.
2. If the operation count has exceeded the threshold, clean up.
3. When cleaning up, set the threshold to 2 * the current size.

That is still amortized, doesn&apos;t have the issue Chris fixed, and is simpler
to reason about and easier to see what is going on.

* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/WeakHashMap.h:
* Source/WTF/wtf/WeakHashSet.h:
* Source/WTF/wtf/WeakListHashSet.h:

Canonical link: <a href="https://commits.webkit.org/266713@main">https://commits.webkit.org/266713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f57c98cc5e397ee0de2e04ecd93199e72c04d413

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14309 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14620 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16047 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13545 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14432 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17133 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14697 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16213 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14486 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16765 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12320 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12897 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19909 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12242 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13397 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13061 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16271 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13600 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13613 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11461 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14352 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12900 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3712 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17236 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14741 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1736 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13458 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3526 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -177,23 +179,23 @@ class ThreadSafeWeakHashSet final {
             });
-            m_maxOperationCountWithoutCleanup = didRemove ? std::max(initialMaxOperationCountWithoutCleanup, m_maxOperationCountWithoutCleanup / 2) : m_maxOperationCountWithoutCleanup * 2;
             m_operationCountSinceLastCleanup = 0;
+            m_maxOperationCountWithoutCleanup = m_map.size() * 2;
```
Should we protect against overflows and use std::numeric_limits<unsigned>::max() when `m_map.size() >= std::numeric_limits<unsigned>::max() / 2` ?

Otherwise we'll wrap against when the size becomes to large and start using small values for `m_maxOperationCountWithoutCleanup` again.

cdumez commented on Source/WTF/wtf/WeakHashMap.h:
```diff
@@ -351,13 +357,12 @@ class WeakHashMap final {
-            m_maxOperationCountWithoutCleanup = didRemove ? std::max(initialMaxOperationCountWithoutCleanup, m_maxOperationCountWithoutCleanup / 2) : m_maxOperationCountWithoutCleanup * 2;
+        if (currentCount > m_maxOperationCountWithoutCleanup) {
+            const_cast<WeakHashMap&>(*this).removeNullReferences();
+            m_maxOperationCountWithoutCleanup = m_map.size() * 2;
```
removeNullReferences() already does this.

cdumez commented on Source/WTF/wtf/WeakHashSet.h:
```diff
@@ -190,19 +194,18 @@ class WeakHashSet final {
-            m_maxOperationCountWithoutCleanup = didRemove ? std::max(initialMaxOperationCountWithoutCleanup, m_maxOperationCountWithoutCleanup / 2) : m_maxOperationCountWithoutCleanup * 2;
+        if (currentCount > m_maxOperationCountWithoutCleanup) {
+            const_cast<WeakHashSet&>(*this).removeNullReferences();
+            m_maxOperationCountWithoutCleanup = m_set.size() * 2;
```
removeNullReferences() already does this.

cdumez commented on Source/WTF/wtf/WeakListHashSet.h:
```diff
@@ -317,19 +321,18 @@ class WeakListHashSet final {
-            m_maxOperationCountWithoutCleanup = didRemove ? std::max(initialMaxOperationCountWithoutCleanup, m_maxOperationCountWithoutCleanup / 2) : m_maxOperationCountWithoutCleanup * 2;
+        if (currentCount > m_maxOperationCountWithoutCleanup) {
+            const_cast<WeakListHashSet&>(*this).removeNullReferences();
+            m_maxOperationCountWithoutCleanup = m_set.size() * 2;
```
removeNullReferences() already does this.

cdumez reviewed (commented):
This is a little simpler indeed.

cdumez commented on Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
```diff
@@ -2897,7 +2897,14 @@ class ObjectAddingAndRemovingItself : public ThreadSafeRefCountedAndCanMakeThrea
-        EXPECT_TRUE(m_set.contains(*this));
+        static size_t i { 0 };
+        i++;
+        if (i == 8) {
```
I think we would have dropped this `EXPECT_TRUE(m_set.contains(*this));` check instead of giving special treatment to `i ==8`

achristensen07 replied:
I think we should make ThreadSafeWeakHashSet::contains to return false if the weak pointer is null, but that is unrelated to the amortization algorithm, so I'm planning to do it in a separate PR.

achristensen07 replied:
Following up in https://github.com/WebKit/WebKit/pull/16512

cdumez reviewed (approved):
LGTM

achristensen07 committed:
Simplify amortized cleanup algorighm for weak hash structures
https://bugs.webkit.org/show_bug.cgi?id=259847
rdar://113419306

Reviewed by Chris Dumez.

In 266594@main Chris fixed an important bug where if you only add to a
ThreadSafeHashSet, you never hit the cleanup condition of the container.
Other containers had this same issue.  This simplifies the algorithm used
to determine whether to clean up the container to just this:

1. Increase the operation count when an add/remove operation happens.
2. If the operation count has exceeded the threshold, clean up.
3. When cleaning up, set the threshold to 2 * the current size.

That is still amortized, doesn't have the issue Chris fixed, and is simpler
to reason about and easier to see what is going on.

* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/WeakHashMap.h:
* Source/WTF/wtf/WeakHashSet.h:
* Source/WTF/wtf/WeakListHashSet.h:

Canonical link: https://commits.webkit.org/266713@main
